tools/wheelmaker.py

```diff
@@ -32,9 +34,54 @@ def commonpath(path1, path2):
     return os.path.sep.join(ret)
 
 
-def escape_filename_segment(segment):
-    """Escapes a filename segment per https://www.python.org/dev/peps/pep-0427/#escaping-and-unicode"""
-    return re.sub(r"[^\w\d.]+", "_", segment, re.UNICODE)
+def normalize_package_name(name):
+    """Normalize a package name according to the Python Packaging User Guide.
+
+    See https://packaging.python.org/en/latest/specifications/name-normalization/
+    """
+    return re.sub(r"[-_.]+", "-", name).lower()
+
+
+def escape_filename_distribution_name(name):
+    """Escape the distribution name component of a filename.
+
+    See https://packaging.python.org/en/latest/specifications/binary-distribution-format/#escaping-and-unicode
+    """
+    return normalize_package_name(name).replace("-", "_")
+
+
+def normalize_pep440(version):
+    """Normalize version according to PEP 440, with fallback for placeholders.
+
+    If there's a placeholder in braces, such as {BUILD_TIMESTAMP},
+    replace it with 0. Such placeholders can be used with stamping, in
+    which case they would have been resolved already by now; if they
+    haven't, we're doing an unstamped build, but we still need to
+    produce a valid version. If such replacements are made, the
+    original version string, sanitized to dot-separated alphanumerics,
+    is appended as a local version segment, so you understand what
+    placeholder was involved.
+
+    If that still doesn't produce a valid version, use version 0 and
+    append the original version string, sanitized to dot-separated
+    alphanumerics, as a local version segment.
+
+    """
+
+    try:
+        return str(packaging.version.Version(version))
+    except packaging.version.InvalidVersion:
+        pass
+
+    sanitized = re.sub(r'[^a-z0-9]+', '.', version.lower()).strip('.')
+    substituted = re.sub(r'\{\w+\}', '0', version)
```

Comment:
I assume this has been taken from somewhere, could you paste a link to the source please?

Reply:
It wasn't copy-pasted, if that's what you're thinking. 

Reference for `version.lower()`: https://peps.python.org/pep-0440/#case-sensitivity
Reference for `re.sub(..., '.', ...)`: https://peps.python.org/pep-0440/#local-version-segments
Reference for only allowing a-z, 0-9 and period: https://peps.python.org/pep-0440/#local-version-identifiers
Reference for removing periods from beginning and end, and collapsing sequences of multiple periods to one: https://peps.python.org/pep-0440/#appendix-b-parsing-version-strings-with-regular-expressions

Substituting Bazel placeholders with "0" is essentially arbitrary, but I took inspiration from https://github.com/bazelbuild/rules_python/pull/1331#issuecomment-1668422845

See also the docstring for the purpose of this: this is Bazel-specific functionality to coerce a non-PEP440 version with a placeholder into a PEP440-compliant version, in the case the build is performed without stamping. We can do whatever we want here, as long as the result is a normalized PEP440 version.

Reply:
Thanks for explaining. I am wondering if some of this should be documented in the code.

Reply:
I see the docstring now. I would prefer code comments in here, but I am fine with the current state as well.